File: java/client/src/main/java/asyncCache/client/MemCacheServerPersistence.java
package asyncCache.client;

import java.io.IOException;
import java.time.Duration;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;

import asyncMemManager.common.di.Persistence;
import okhttp3.ConnectionPool;
import okhttp3.OkHttpClient;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.Path;

public class MemCacheServerPersistence implements Persistence{
	private AsyncCachingREST restClient;
	
	public MemCacheServerPersistence(String asyncCachingUrl) {
		ConnectionPool pool = new ConnectionPool(5, 1, TimeUnit.MINUTES);

		OkHttpClient client = new OkHttpClient.Builder()
									  .readTimeout(Duration.ofSeconds(20))
		                              .connectionPool(pool)
		                              .build();
		
		Retrofit retrofit = new Retrofit.Builder()
	            .client(client)
			    .baseUrl(asyncCachingUrl)
			    .addConverterFactory(GsonConverterFactory.create())
			    .build();
		
		this.restClient = retrofit.create(AsyncCachingREST.class);
	}
	
	@Override
	public CompletableFuture<Void> store(UUID key, String data, long expectedDuration) {
		CompletableFuture<Void> res = new CompletableFuture<Void>();
		Call<Void> call = this.restClient.store(key, data, expectedDuration);
		call.enqueue(new Callback<Void>() {
			@Override
		    public void onResponse(Call<Void> call, Response<Void> response) {
		        res.complete(null);  
		    }

		    @Override
		    public void onFailure(Call<Void> call, Throwable t) {
		        // Log error here since request failed
		    	res.completeExceptionally(t);
		    }
		});
		return res;
	}

	@Override
	public String retrieve(UUID key) {
		try {
			return this.restClient.retrieve(key).execute().body();
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		return null;
	}

	@Override
	public void remove(UUID key) {
		try {
			this.restClient.remove(key).execute().body();
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}	
	}
	
	private static interface AsyncCachingREST
	{
		@POST("/cache/store/{key}/{expectedDuration}")
		public Call<Void> store(@Path("key")UUID key, @Body String data, @Path("expectedDuration") long expectedDuration) ;

		@GET("/cache/retrieve/{key}")
		public Call<String> retrieve(@Path("key") UUID key);

		@GET("/cache/remove/{key}")
		public Call<Void> remove(@Path("key") UUID key);
	}

}

File: .net/Client/MemCacheServerPersistence.cs
namespace AsyncMemManager.Client
{
    using System;
    using System.Threading.Tasks;
    using RestSharp;

    public class MemCacheServerPersistence : DI.IPersistence
    {
        private RestClient client;

        public MemCacheServerPersistence(string asyncCachingUrl)
        {
            this.client = new RestClient(asyncCachingUrl);
            this.client.ThrowOnAnyError = true;
        }

		public void Store(Guid key, string data, long expectedDuration)
        {
            var request = new RestRequest("/cache/{key}/{expectedDuration}")
                            .AddUrlSegment("key", key)
                            .AddUrlSegment("expectedDuration", expectedDuration)
                            .AddParameter("text/plain", data, ParameterType.RequestBody);
            var res = client.Post(request);
            if ((int)res.StatusCode != 200)
            {
                Console.WriteLine("Failed to store");
            }
        }
		
		/**
		* retrieve and remove data from storage
		* @param key
		* @return
		*/
		public string Retrieve(Guid key)
        {
            var request = new RestRequest("/cache/{key}")
                .AddUrlSegment("key", key);
            var response = client.Get<string>(request);            
            return response.Content;
        }
		
		/**
		* remove data from storage.
		* @param key
		*/
		public void Remove(Guid key)
        {
            var request = new RestRequest("/cache/{key}")
                .AddUrlSegment("key", key);
            this.client.Delete(request);
        }
    }
}